src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const states = [];
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        newState = { ...newState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+
+      default:
+        break;
+    }
+
+    states.push(newState);
```

Comment:
To avoid unnecessary mutations and to follow the checklist, push the copy of the state object to the `states` array at the end of each loop cycle, but outside of the `switch` block, after the current action is processed. This ensures that the `currentState` variable is only updated once per iteration.